formats/json/jvmMain/src/kotlinx/serialization/json/internal/CharArrayPool.kt

```diff
@@ -0,0 +1,24 @@
+package kotlinx.serialization.json.internal
+
+import java.util.concurrent.*
+
+internal object CharArrayPool {
+    private val arrays = ArrayDeque<CharArray>()
+    private var charsTotal = 0
+    private val MAX_CHARS_IN_POOL = runCatching {
+        System.getProperty("kotlinx.serialization.json.pool.size").toIntOrNull()
+    }.getOrNull() ?: 1024 * 1024 // 2 MB seems to be a reasonable contraint
+
+    public fun take(): CharArray {
+        val candidate = synchronized(this) {
+            arrays.lastOrNull()?.also { charsTotal -= it.size }
+        }
+        return candidate ?: CharArray(128)
```

Comment:
More or less. I left a comment

>It helps eliminating new allocations in StringBuilder creation

Not really, otherwise we could have cached `StringBuilder` instances and invoke `setLength(0)`. Regarding this, I'll leave a comment in `JsonStringBuilder`